kate/recovery/src/com.rs

```diff
@@ -1,39 +1,106 @@
+) -> Result<HashMap<u16, HashMap<u16, Cell>>> {
+	let mut result: HashMap<u16, HashMap<u16, Cell>> = HashMap::new();
+	for cell in cells {
+		let row = cell.row;
+		let col = cell.col;
+		if row as usize > dimensions.rows * 2 || col as usize > dimensions.cols {
+			return Err(anyhow!("Invalid cell (col {}, row {})", col, row));
+		}
+		let cells = result.entry(col).or_insert_with(HashMap::new);
+		if cells.insert(cell.row, cell).is_some() {
+			return Err(anyhow!("Duplicate cell found"));
+		}
+	}
+	Ok(result)
+}
+
+/// Reconstructs app extrinsics from extrinsics layout and data.
+/// If app_id is `None`, all extrinsics are reconstructed.
+/// If app_id is provided, only related extrinsics are reconstructed.
+/// Only related data cells needs to be in matrix (unrelated columns can be empty).
+///
+///
+/// # Arguments
+///
+/// * `layout` - Extrinsics layout, vector of app_id and size in chunks pairs
+/// * `dimensions` - Matrix dimensions
+/// * `cells` - Cells from required columns, at least 50% cells per column
+/// * `app_id` - Optional application id
 pub fn reconstruct_app_extrinsics(
-	xt_layout: Vec<(u32, u32)>,
-	columns: Vec<Vec<Cell>>,
-	column_length: usize,
-	chunk_size: usize,
-) -> Vec<(u32, Vec<u8>)> {
-	let reconstructed = columns
-		.iter()
-		.map(|cells| reconstruct_column(column_length * 2, cells).unwrap())
-		.collect::<Vec<_>>();
-	let scalars = reconstructed
-		.iter()
-		.flat_map(|e| e.iter().flat_map(|e| e.to_bytes()).collect::<Vec<_>>())
-		.collect::<Vec<_>>();
+	layout: &[(u32, u32)],
+	dimensions: &MatrixDimensions,
+	cells: Vec<Cell>,
+	app_id: Option<u32>,
+) -> Result<Vec<(u32, Vec<u8>)>> {
+	let mut column_numbers: Vec<u16> = vec![];
+	let mut data: Vec<u8> = vec![];
+	let cells_map = map_cells(dimensions, cells)?;
+	for column_number in 0..dimensions.cols as u16 {
+		match cells_map.get(&column_number) {
+			None => data.extend(vec![0; dimensions.rows * dimensions.chunk_size]),
+			Some(column_cells) => {
+				if column_cells.len() < dimensions.rows {
+					return Err(anyhow!(
+						"Column {} contains less than half rows",
+						column_number
+					));
+				}
+				let cells = column_cells.values().cloned().collect::<Vec<_>>();
+				let scalars =
+					reconstruct_column(dimensions.rows * 2, &cells).map_err(|err| anyhow!(err))?;
+				let column_data = scalars.iter().flat_map(|e| e.to_bytes());
+				column_numbers.push(column_number);
+				data.extend(column_data);
+			},
+		}
+	}
 
-	unflatten_padded_data(xt_layout, scalars, chunk_size)
+	let ranges = data_ranges(layout)
+		.into_iter()
+		.filter(|(id, _)| app_id.is_none() || Some(*id) == app_id)
+		.collect::<Vec<_>>();
+	Ok(unflatten_padded_data(ranges, data, dimensions.chunk_size))
 }
 
 fn trim_to_chunk_data(chunk: &[u8]) -> [u8; DATA_CHUNK_SIZE] {
 	assert!(DATA_CHUNK_SIZE < chunk.len(), "Cannot trim to bigger size!");
 	chunk[0..DATA_CHUNK_SIZE].try_into().unwrap()
 }
 
+pub fn data_ranges(layout: &[(u32, u32)]) -> Vec<(u32, Range<usize>)> {
```

Comment:
This is obvious what it does, but a comment couldn't hurt.

Reply:
Done.